File: 6lab/cs/Basket.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using MusicStore;

namespace MusicStore
{
    public class Basket
    {
        // Константы
        private const int MAX_PRODUCTS = 100; // максимальное количество товаров в корзине

        // Поля класса
        private Product[] products;           // массив товаров
        private int[] quantities;             // массив количеств
        private int productCount;             // текущее количество товаров в корзине

        // Конструктор
        public Basket()
        {
            productCount = 0;
            products = new Product[MAX_PRODUCTS];
            quantities = new int[MAX_PRODUCTS];

            // Инициализация массивов
            for (int i = 0; i < MAX_PRODUCTS; i++)
            {
                products[i] = null;
                quantities[i] = 0;
            }
        }

        // Метод для расчета общей стоимости всех товаров в корзине
        public double GetBasketCost()
        {
            double totalCost = 0.0;
            // Проход по всем товарам в корзине
            for (int i = 0; i < productCount; i++)
            {
                if (products[i] != null)
                {// Добавляем стоимость товара: цена × количество
                    totalCost += products[i].GetPrice() * quantities[i];
                }
            }

            return totalCost;
        }

        // Метод для получения общего количества всех товаров в корзине
        public uint GetBasketCount()
        {
            uint totalCount = 0;

            for (int i = 0; i < productCount; i++)
            {
                totalCount += (uint)quantities[i];
            }

            return totalCount;
        }

        // Метод отображения содержимого корзины
        public void DisplayBasket()
        {
            Console.WriteLine("\n=== СОДЕРЖИМОЕ КОРЗИНЫ ===");

            if (productCount == 0)
            {
                Console.WriteLine("Корзина пуста!");
                Console.WriteLine("=========================");
                return;
            }

            Console.WriteLine($"{"№",-3} {"Артикул",-10} {"Название",-20} {"Цена",-8} {"Кол-во",-10} {"Сумма",-12}");
            Console.WriteLine(new string('-', 70));

            double totalCost = 0.0;// Переменная для общей стоимости

            for (int i = 0; i < productCount; i++)
            {
                if (products[i] != null)
                {// Расчет стоимости позиции: цена × количество
                    double itemTotal = products[i].GetPrice() * quantities[i];
                    totalCost += itemTotal;// Добавление к общей сумме
                    // Обработка слишком длинных названий (обрезаем если > 18 символов)
                    string productName = products[i].GetName();
                    if (productName.Length > 18)
                    {
                        productName = productName.Substring(0, 18) + "..";
                    }
                    // Вывод строки с информацией о товаре
                    Console.WriteLine($"{i + 1,-3} {products[i].GetArticle(),-10} {productName,-20} " +
                                      $"{products[i].GetPrice(),-8} {quantities[i],-10} {itemTotal,-12:F2}");
                }
            }

            Console.WriteLine(new string('-', 70));
            Console.WriteLine($"ИТОГО: {totalCost:F2} руб.");
            Console.WriteLine($"Общее количество товаров: {GetBasketCount()}");
            Console.WriteLine("=========================");
        }

        // Метод добавления товара в корзину
        public void AddBasket(Product product, int quantity = 1)
        {
            // Проверка на переполнение корзины
            if (productCount >= MAX_PRODUCTS)
            {
                Console.WriteLine($"Корзина переполнена! Максимум {MAX_PRODUCTS} товаров.");
                return;
            }

            // Проверка корректности количества
            if (quantity <= 0)
            {
                Console.WriteLine("Количество должно быть положительным числом!");
                return;
            }

            // Проверяем, есть ли уже такой товар в корзине
            int existingIndex = FindProductIndex(product.GetArticle());

            // Если товар найден, увеличиваем количество
            if (existingIndex != -1)
            {
                quantities[existingIndex] += quantity;
                Console.WriteLine($"Количество товара '{product.GetName()}' увеличено на {quantity}. " +
                                 $"Теперь: {quantities[existingIndex]} шт.");
            }
            else
            {
                // Добавляем новый товар
                products[productCount] = product;
                quantities[productCount] = quantity;
                productCount++;
                Console.WriteLine($"Товар '{product.GetName()}' добавлен в корзину в количестве {quantity} шт.");
            }
        }

        // Метод удаления товара из корзины по индексу
        public void DeleteBasket(int productIndex)
        {
            // Проверка корректности индекса (пользователь вводит с 1, а не с 0)
            if (productIndex < 1 || productIndex > productCount)
            {
                Console.WriteLine("Неверный номер товара!");
                return;
            }

            int actualIndex = productIndex - 1;// Преобразуем в индекс массива
            string productName = products[actualIndex].GetName();//название

            // Сдвигаем элементы массива влево, начиная с удаляемого элемента
            for (int i = actualIndex; i < productCount - 1; i++)
            {
                products[i] = products[i + 1];  // сдвиг указателей на товары
                quantities[i] = quantities[i + 1];  // сдвиг количеств
            }

            // Очищаем последнюю ячейку
            products[productCount - 1] = null;
            quantities[productCount - 1] = 0;
            productCount--;

            Console.WriteLine($"Товар '{productName}' удален из корзины.");
        }

        // Метод обновления количества товара по индексу
        public int UpdateQuantity(int productIndex, int newQuantity)
        {
            if (productIndex < 1 || productIndex > productCount)
            {
                Console.WriteLine("Неверный номер товара!");
                return -1;
            }

            if (newQuantity <= 0)
            {
                Console.WriteLine("Количество должно быть положительным числом!");
                return -1;
            }

            int actualIndex = productIndex - 1;
            int oldQuantity = quantities[actualIndex];
            quantities[actualIndex] = newQuantity;

            Console.WriteLine($"Количество товара '{products[actualIndex].GetName()}' " +
                             $"изменено с {oldQuantity} на {newQuantity} шт.");

            return newQuantity;
        }

        // Метод проверки, пуста ли корзина
        public bool IsEmpty()
        {
            return productCount == 0;
        }

        // Метод полной очистки корзины
        public void ClearBasket()
        {
            for (int i = 0; i < productCount; i++)
            {
                products[i] = null;
                quantities[i] = 0;
            }
            productCount = 0;
        }

        // Метод поиска индекса товара в корзине по артикулу
        public int FindProductIndex(int article)
        {
            for (int i = 0; i < productCount; i++)
            {
                if (products[i] != null && products[i].GetArticle() == article)
                {
                    return i;
                }
            }
            return -1;
        }

        // Метод получения количества различных товаров в корзине
        public int GetProductCount()
        {
            return productCount;
        }

        // Метод получения товара по индексу с проверкой границ массива
        public Product GetProduct(int index)
        {
            if (index < 0 || index >= productCount)
            {
                return null;
            }
            return products[index];
        }

        // Метод получения количества товара по индексу с проверкой границ массива
        public int GetQuantity(int index)
        {
            if (index < 0 || index >= productCount)
            {
                return 0;
            }
            return quantities[index];
        }

        // Свойства для доступа к полям
        public int ProductCount
        {
            get { return productCount; }
        }


        ~Basket()
        {
            ClearBasket();
        }
    }
}

File: cpp/src/Basket.cpp
#include "Basket.h"
#include <iostream>
#include <iomanip>//форматирование текста и таблиц

using namespace std;

Basket::Basket() : productCount(0) {// инициализация счетчика товаров нулем
    // Инициализация массивов корзины
    for (int i = 0; i < MAX_PRODUCTS; i++) {
        products[i] = nullptr;// обнуляем все указатели на товары
        quantities[i] = 0;// обнуляем все количества
    }
}
// Деструктор класса Basket - очищает корзину при уничтожении объекта
Basket::~Basket() {
    ClearBasket(); //метод очистки корзины
}

// Метод для расчета общей стоимости всех товаров в корзине
double Basket::GetBasketCost() const {
    double totalCost = 0.0;// переменная для накопления общей стоимости
    // Проходим по всем товарам в корзине
    for (int i = 0; i < productCount; i++) {
        // Проверяем что указатель на товар не нулевой
        if (products[i] != nullptr) {
            // Суммируем: цена товара * количество
            totalCost += products[i]->getPrice() * quantities[i];
        }
    }
    return totalCost;
}

// Метод для получения общего количества всех товаров в корзине
unsigned int Basket::GetBasketCount() const {
    unsigned int totalCount = 0;// переменная для подсчета общего количества
    for (int i = 0; i < productCount; i++) {// добавляем количество каждого товара
        totalCount += quantities[i];
    }
    return totalCount;
}

void Basket::DisplayBasket() const {
    cout << "\n=== СОДЕРЖИМОЕ КОРЗИНЫ ===" << endl;
    if (productCount == 0) {
        cout << "Корзина пуста!" << endl;
        cout << "=========================" << endl;
        return;
    }

    cout << left << setw(3) << "№" << setw(10) << "Артикул"
        << setw(20) << "Название" << setw(8) << "Цена"
        << setw(10) << "Кол-во" << setw(12) << "Сумма" << endl;
    cout << "------------------------------------------------------------" << endl;

    double totalCost = 0.0;//общая стоимость

    // Вывод информации о каждом товаре в корзине
    for (int i = 0; i < productCount; i++) {
        // Проверяем что товар существует
        if (products[i] != nullptr) {
            double itemTotal = products[i]->getPrice() * quantities[i];
            totalCost += itemTotal;// добавляем к общей стоимости
            // Вывод информации о товаре с форматированием
            cout << left << setw(3) << (i + 1)
                << setw(10) << products[i]->getArticle()
                << setw(20) << (products[i]->getName().length() > 18 ?
                    products[i]->getName().substr(0, 18) + ".." :
                    products[i]->getName())
                << setw(8) << products[i]->getPrice()
                << setw(10) << quantities[i]
                << setw(12) << itemTotal << endl;
        }
    }

    cout << "------------------------------------------------------------" << endl;
    cout << "ИТОГО: " << totalCost << " руб." << endl;
    cout << "Общее количество товаров: " << GetBasketCount() << endl;
    cout << "=========================" << endl;
}

// Добавление товара в корзину
void Basket::AddBasket(Product* product, int quantity) {
    // Проверка на переполнение корзины
    if (productCount >= MAX_PRODUCTS) {
        cout << "Корзина переполнена! Максимум " << MAX_PRODUCTS << " товаров." << endl;
        return;
    }
    // Проверка корректности количества
    if (quantity <= 0) {
        cout << "Количество должно быть положительным числом!" << endl;
        return;
    }

    // Проверяем, есть ли уже такой товар в корзине
    int existingIndex = FindProductIndex(product->getArticle());
    //если товар найден, увеличиваем количество
    if (existingIndex != -1) {
        quantities[existingIndex] += quantity;
        cout << "Количество товара '" << product->getName()
            << "' увеличено на " << quantity << ". Теперь: "
            << quantities[existingIndex] << " шт." << endl;
    }
    else {
        // Добавляем новый товар
        products[productCount] = product; // сохраняем указатель на товар
        quantities[productCount] = quantity;// сохраняем количество
        productCount++;
        cout << "Товар '" << product->getName()
            << "' добавлен в корзину в количестве " << quantity << " шт." << endl;
    }
}

// Удаление товара из корзины по индексу
void Basket::DeleteBasket(int productIndex) {
    // Проверка корректности индекса (пользователь вводит с 1, а не с 0)
    if (productIndex < 1 || productIndex > productCount) {
        cout << "Неверный номер товара!" << endl;
        return;
    }

    int actualIndex = productIndex - 1;// сохраняем старое количество
    string productName = products[actualIndex]->getName();// сохраняем имя для сообщения

    // Сдвигаем элементы массива влево, начиная с удаляемого элемента
    for (int i = actualIndex; i < productCount - 1; i++) {
        products[i] = products[i + 1];// сдвиг указателей на товары
        quantities[i] = quantities[i + 1];// сдвиг количеств
    }

    products[productCount - 1] = nullptr;// Очищаем последнюю ячейку (которая теперь дублируется)
    quantities[productCount - 1] = 0;
    productCount--;// уменьшаем счетчик товаров

    cout << "Товар '" << productName << "' удален из корзины." << endl;
}

// Обновление количества товара по индексу
int Basket::UpdateQuantity(int productIndex, int newQuantity) {
    if (productIndex < 1 || productIndex > productCount) {
        // Проверка корректности индекса
        cout << "Неверный номер товара!" << endl;
        return -1;
    }

    if (newQuantity <= 0) {
        cout << "Количество должно быть положительным числом!" << endl;
        return -1;
    }

    int actualIndex = productIndex - 1;// Преобразуем пользовательский индекс в программный
    int oldQuantity = quantities[actualIndex];// сохраняем старое количество
    quantities[actualIndex] = newQuantity;// устанавливаем новое количество

    cout << "Количество товара '" << products[actualIndex]->getName()
        << "' изменено с " << oldQuantity << " на " << newQuantity << " шт." << endl;
    return newQuantity;// возвращаем новое количество
}

// Проверка, пуста ли корзина
bool Basket::IsEmpty() const {
    return productCount == 0;// true если товаров нет, false если есть
}

// Полная очистка корзины
void Basket::ClearBasket() {
    // Проходим по всем занятым ячейкам корзины
    for (int i = 0; i < productCount; i++) {
        // Не удаляем сами продукты, только обнуляем указатели
        products[i] = nullptr;
        quantities[i] = 0;
    }
    productCount = 0;// сбрасываем счетчик товаров
}

// Поиск индекса товара в корзине по артикулу
int Basket::FindProductIndex(int article) const {
    // Проходим по всем товарам в корзине
    for (int i = 0; i < productCount; i++) {
        // Проверяем что указатель не нулевой и артикул совпадает
        if (products[i] != nullptr && products[i]->getArticle() == article) {
            return i; // возвращаем индекс найденного товара
        }
    }
    return -1;
}

//нужны для безопасного доступа к данным из корзины,защищая от ошибок
// Геттер для получения товара по индексу
Product* Basket::getProduct(int index) const {
    // Проверка корректности индекса
    if (index < 0 || index >= productCount) {
        return nullptr;// возвращаем nullptr при неверном индексе
    }
    return products[index];// возвращаем указатель на товар
}

// Геттер для получения количества товара по индексу
int Basket::getQuantity(int index) const {
    // Проверка корректности индекса
    if (index < 0 || index >= productCount) {
        return 0;// возвращаем 0 при неверном индексе
    }
    return quantities[index];// возвращаем количество товара
}